node/narwhal/src/helpers/storage.rs

```diff
@@ -817,22 +817,258 @@ pub mod tests {
+        fn next_u32(&mut self) -> u32 {
+            self.0.next_u32()
+        }
+
+        fn next_u64(&mut self) -> u64 {
+            self.0.next_u64()
+        }
+
+        fn fill_bytes(&mut self, dest: &mut [u8]) {
+            self.0.fill_bytes(dest);
+        }
+
+        fn try_fill_bytes(&mut self, dest: &mut [u8]) -> std::result::Result<(), Error> {
+            self.0.try_fill_bytes(dest)
+        }
+    }
+
+    impl CryptoRng for CryptoTestRng {}
+
+    #[derive(Debug, Clone)]
+    pub struct AnyTransmission(pub Transmission<CurrentNetwork>);
+
+    impl Arbitrary for AnyTransmission {
+        type Parameters = ();
+        type Strategy = BoxedStrategy<AnyTransmission>;
+
+        fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+            any_transmission().prop_map(AnyTransmission).boxed()
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct AnyTransmissionID(pub TransmissionID<CurrentNetwork>);
+
+    impl Arbitrary for AnyTransmissionID {
+        type Parameters = ();
+        type Strategy = BoxedStrategy<AnyTransmissionID>;
+
+        fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+            any_transmission_id().prop_map(AnyTransmissionID).boxed()
+        }
+    }
+
+    fn any_transmission() -> BoxedStrategy<Transmission<CurrentNetwork>> {
+        prop_oneof![
+            (collection::vec(any::<u8>(), 512..=512))
+                .prop_map(|bytes| Transmission::Solution(Data::Buffer(Bytes::from(bytes)))),
+            (collection::vec(any::<u8>(), 2048..=2048))
+                .prop_map(|bytes| Transmission::Transaction(Data::Buffer(Bytes::from(bytes)))),
+        ]
+        .boxed()
+    }
+
+    fn any_puzzle_commitment() -> BoxedStrategy<PuzzleCommitment<CurrentNetwork>> {
+        Just(0).prop_perturb(|_, rng| PuzzleCommitment::from_g1_affine(CryptoTestRng(rng).gen())).boxed()
+    }
+
+    fn any_transaction_id() -> BoxedStrategy<<CurrentNetwork as Network>::TransactionID> {
+        Just(0)
+            .prop_perturb(|_, rng| {
+                <CurrentNetwork as Network>::TransactionID::from(Field::rand(&mut CryptoTestRng(rng)))
+            })
+            .boxed()
+    }
+
+    fn any_transmission_id() -> BoxedStrategy<TransmissionID<CurrentNetwork>> {
+        prop_oneof![
+            any_transaction_id().prop_map(TransmissionID::Transaction),
+            any_puzzle_commitment().prop_map(TransmissionID::Solution),
+        ]
+        .boxed()
+    }
+
+    impl ValidatorSet {
+        fn sign_batch_header<R: Rng + CryptoRng>(
+            &self,
+            batch_header: &BatchHeader<CurrentNetwork>,
+            rng: &mut R,
+        ) -> IndexMap<Signature<CurrentNetwork>, i64> {
+            let mut signatures = IndexMap::with_capacity(self.0.len());
+            for validator in self.0.iter() {
+                let private_key = validator.account.private_key();
+                let timestamp = time::OffsetDateTime::now_utc().unix_timestamp();
```

Comment:
perhaps it would be preferable to obtain the timestamp just once before the loop? or do we want all of the validators to have a (very) slightly different one?

Reply:
As this fn is strictly for testing purposes, the slightly different signature timestamp for each validator is the one that reflects the target system behavior better. That said, proptests would be a better fit for testing the `BatchCertificate` and related logic for any combinations of timestamp variation.